File: wlr/OGUR/OGUR/Creatures/Stats.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using OGUR.Util;

namespace OGUR.Creatures
{
    public class Stats
    {
        public const float DefaultMoveSpeed = Sprites.SpriteInfo.Height;
        public const float DefaultCoolDown = 18;
        public const float DefaultRegenRate = 1;

        private readonly Dictionary<string, float> m_stats = new Dictionary<string, float>();
        private readonly List<StatBuff> m_buffs = new List<StatBuff>(); 

        public Stats(Stats target)
        {
            m_stats = new Dictionary<string, float>(target.m_stats);
        }
        public Stats
            (
                float health,
                float mana,
                float strength,
                float wisdom,
                float defense,
                float luck,
                float age,
                float weightInLbs,
                float heightInFeet,
                float moveCoolDown = DefaultCoolDown,
                float regenRate = DefaultRegenRate
            )
        {
            Setup(new List<float>{health, mana, strength, wisdom, defense,luck, age,weightInLbs, heightInFeet, moveCoolDown,0,regenRate});
        }

        private void Setup(List<float> stats)
        {
            for (var ii = 0; ii < stats.Count; ii++)
            {
                m_stats.Add(StatType.Values[ii], stats[ii]);
            }
        }
        private float statSum = 0;
        public float Get(string stat)
        {
            if (m_buffs != null)
            {
                if (!m_buffs.Contains(null))
                {
                    statSum = 0;
                    for (int ii = 0; ii < m_buffs.Count(); ii++)
                    {
                        if (m_buffs[ii].Stat == stat)
                        {
                            statSum += m_buffs[ii].Amount;
                        }
                    }
                    return GetRaw(stat) + statSum;
                }
            }
            return GetRaw(stat);
        }

        public float GetRaw(string stat)
        {
            return m_stats[stat];
        }

        public float Set(string stat, float value)
        {
            return m_stats[stat] = value;
        }

        public void AddBuff(StatBuff buff)
        {
            if(m_buffs.Contains(buff))
            {
                m_buffs.Remove(buff);
                return;
            }
            m_buffs.Add(buff);
        }

        private readonly List<float> deltas = new List<float>();
        public List<float> GetDeltas(Stats stats)
        {
            deltas.Clear();
            foreach (var stat in StatType.Values)
            {
                deltas.Add(m_stats[stat] - stats.m_stats[stat]);
            }
            return deltas;
        }

        public Stats GetLevelBonuses(int level)
        {
            var result = new Stats(this);
            foreach(var stat in result.m_stats.Keys)
            {
                result.m_stats[stat] += result.m_stats[stat]*level;
            }
            return result;
        }

        public float GetBonus(int level, string stat)
        {
            return m_stats[stat]*level;
        }

        public float GetSum()
        {
            return m_stats.Keys.Where(o=>o!=StatType.HEALTH&&o!=StatType.MOVE_COOL_DOWN&&o!=StatType.REGEN).Sum(stat => m_stats[stat]);
        }

        float hash = 0;
        public override int GetHashCode()
        {
            hash = 0;
            foreach(var key in m_stats.Keys)
            {
                hash += m_stats[key];
            }
            return hash.GetHashCode();
        }
    }
}

File: proj/src/aigilas/creatures/Stats.java
package aigilas.creatures;

import spx.core.Settings;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Stats {
    public static float DefaultMoveDistance = Settings.Get().spriteHeight;

    private HashMap<StatType, Float> _stats = new HashMap<>();
    private final List<StatBuff> _buffs = new ArrayList<StatBuff>();

    public Stats(Stats target) {
        _stats = new HashMap<StatType, Float>(target._stats);
    }

    public Stats(float health, float mana, float strength, float wisdom, float defense, float luck, float age, float weightInLbs, float heightInFeet, float moveCoolDown, float regenRate) {

        Setup(health, mana, strength, wisdom, defense, luck, age, weightInLbs, heightInFeet, moveCoolDown, 0, regenRate);
    }

    public Stats(float health, float mana, float strength, float wisdom, float defense, float luck, float age, float weightInLbs, float heightInFeet, float moveCoolDown) {
        Setup(health, mana, strength, wisdom, defense, luck, age, weightInLbs, heightInFeet, moveCoolDown, 0, Settings.Get().defaultRegen);
    }

    public Stats(float health, float mana, float strength, float wisdom, float defense, float luck, float age, float weightInLbs, float heightInFeet) {
        Setup(health, mana, strength, wisdom, defense, luck, age, weightInLbs, heightInFeet, Settings.Get().defaultSpeed, 0, Settings.Get().defaultRegen);

    }

    private void Setup(float... list) {
        for (int ii = 0; ii < list.length; ii++) {
            _stats.put(StatType.values()[ii], list[ii]);
        }
    }

    private float statSum = 0;

    public float Get(StatType stat) {
        if (_buffs != null) {
            if (!_buffs.contains(null)) {
                statSum = 0;
                for (int ii = 0; ii < _buffs.size(); ii++) {
                    if (_buffs.get(ii).Stat == stat) {
                        statSum += _buffs.get(ii).Amount;
                    }
                }
                return GetRaw(stat) + statSum;
            }
        }
        return GetRaw(stat);
    }

    public float GetRaw(StatType stat) {
        return _stats.get(stat);
    }

    public float Set(StatType stat, float value) {
        return _stats.put(stat, value);
    }

    public void AddBuff(StatBuff buff) {
        if (_buffs.contains(buff)) {
            _buffs.remove(buff);
            return;
        }
        _buffs.add(buff);
    }

    private final List<Float> deltas = new ArrayList<Float>();

    public List<Float> GetDeltas(Stats stats) {
        deltas.clear();
        for (StatType stat : StatType.values()) {
            deltas.add(_stats.get(stat) - stats._stats.get(stat));
        }
        return deltas;
    }

    public Stats GetLevelBonuses(int level) {
        Stats result = new Stats(this);
        for (StatType stat : result._stats.keySet()) {
            result._stats.put(stat, _stats.get(stat) + result._stats.get(stat) * level);
        }
        return result;
    }

    public float GetBonus(int level, StatType stat) {
        return _stats.get(stat) * level;
    }

    public float GetSum() {
        float result = 0;
        for (StatType key : _stats.keySet()) {
            if (key != StatType.HEALTH && key != StatType.MOVE_COOL_DOWN && key != StatType.REGEN) {
                result += _stats.get(key);
            }
        }
        return result;
    }

    Float hash = 0f;

    @Override
    public int hashCode() {
        hash = 0f;
        for (StatType key : _stats.keySet()) {
            hash += _stats.get(key);
        }
        return hash.hashCode();
    }
}
